pkg/config/expand.go

```diff
@@ -412,6 +412,59 @@ type varContext struct {
 	blueprint  Blueprint
 }
 
+/*
+A module reference is made by the use keyword and is subject to IGC constraints
+of references (ordering, explicitness). It has the following fields:
+  - ID: a module ID
+  - GroupID: the deployment group in which the module is *expected* to be found
+  - SourceGroupID: the deployment group from which the reference is made
+  - ExplicitInterGroup: a boolean value indicating whether the user made a
+    reference that superficially appears to be intergroup (i.e. they used
+    an explicit GroupID that differs from the context's GroupID)
+*/
+type modReference struct {
+	ID                 string
+	GroupID            string
+	SourceGroupID      string
+	ExplicitInterGroup bool
+}
+
+/*
+This function performs only the most rudimentary conversion of an input
+string into a modReference struct as defined above. An input string consists of
+1 or 2 fields separated by periods. An error will be returned if there are not
+1 or 2 fields or if either field is the empty string. This function does not
+ensure the existence of the module!
+*/
+func (dg *DeploymentGroup) identifyModuleByReference(yamlReference string) (modReference, error) {
```

Comment:
I think it is possible to imagine a future where there is a "reference" interface

```
type Reference interface {
  validate() error
}
```

and both `modReference` and `varReference` implement that interface. I'd like to support that possibility without requiring it right now.

Reply:
Deleted a comment because I was addressing wrong func. I like attaching this functionality to the DeploymentGroup type because it emphasizes (to my eyes) that the resolution of a YAML string to a reference is a very strong property of the group in which it was made. It's not just another parameter. That said, my read of Go's support for struct methods is that it leads most naturally to implementing interfaces. e.g. I could easily imagine an interface such as

```go

type Reference interface {
  validate([]DeploymentGroups) error
}
```

it is less obvious to me that I could identify a single function name for identifyModule and identifyVar unless we really want to merge the implementations for each. Right now, I don't think that's justified.

Reply:
I have modified this func so that it is not attached to the `DeploymentGroup` type.

Reply:
I agree with your conclusion about validate. For me it would make sense to attach to the deployment group if it would inspect its own data and return a list of module references. The fact you are passing in the reference string makes it feel more like a static method.

Thanks for updating.